python/tvm/dlight/gpu/gemv.py

```diff
@@ -195,106 +201,255 @@ def sch_inner_reduction(  # pylint: disable=too-many-arguments
+                    block_or_loop=sch.get_loops(V_shared)[-4], ann_key="pragma_vectorize", ann_val=1
+                )
+
+            # Schedule epilogue
+            if epilogue_info is not None:
+                epilogue = epilogue_info.block_rv
+                if is_broadcast_epilogue(sch, block, epilogue):
+                    sch.reverse_compute_at(epilogue, bx)
+                    sch.set_scope(block, 0, "shared")
+                    _, _, *s = sch.get_loops(epilogue)  # pylint: disable=invalid-name
+                    _, tx = sch.split(sch.fuse(*s), factors=[None, TX])
+                    sch.bind(tx, "threadIdx.x")
+                else:
+                    sch.reverse_compute_at(epilogue, bx, preserve_unit_loops=True)
+                    ts_tile_s = sch.fuse(*sch.get_loops(epilogue)[1:])
+                    ts_tile_s = sch.get_loops(epilogue)[-1]
+                    ts, tile_s = sch.split(ts_tile_s, factors=[TS, None], preserve_unit_iters=True)
+                    sch.bind(ts, TAG_S)
+                    sch.set_scope(block, 0, "local")
+            # pylint: enable=invalid-name
+            return sch
 
         def get_extent(loop_rv: tir.schedule.LoopRV):
             loop: tir.For = sch.get(loop_rv)
-            return loop.extent.value if isinstance(loop.extent, tir.IntImm) else 1
+            return loop.extent.value if isinstance(loop.extent, tir.IntImm) else loop.extent
 
         # Specify the `len_tx` and `len_ty` according to the loop extent
-        len_s, len_r = get_extent(s), get_extent(r)
-        if len_r >= 4096 and len_r % 128 == 0:
-            len_tx = 128
-        elif 1024 < len_r <= 2048 and len_r % 64 == 0:
-            len_tx = 64
-        else:
-            len_tx = 32
-
-        if len_s >= 4096:
-            len_ty = 8
-        else:
-            len_ty = min(len_s, 4)
-
-        # Use `split_k` to prevent too large shared memory usage
-        split_k: int = 4
-
-        _, tx = sch.split(r, [None, len_tx], preserve_unit_iters=True)
-        # Schedule the RF block
-        rf = sch.rfactor(tx, 0)
-        batch, bx, r, tx, _ = sch.get_loops(rf)
-        sch.reorder(bx, tx, r)
-        ro, ri = sch.split(r, [split_k, None], preserve_unit_iters=True)
-        bx, ty = sch.split(bx, [None, len_ty], preserve_unit_iters=True)
-
-        sch.bind(batch, "blockIdx.y")
-        sch.bind(bx, "blockIdx.x")
-        sch.bind(ty, "threadIdx.y")
-        sch.bind(tx, "threadIdx.x")
-        sch.annotate(ro, "pragma_auto_unroll_max_step", unroll_number)
-        sch.annotate(ro, "pragma_unroll_explicit", 1)
-
+        batch, s, r, c = sch.get_loops(block=block)
+        len_batch, len_s, len_r, len_c = (
+            get_extent(batch),
+            get_extent(s),
+            get_extent(r),
+            get_extent(c),
+        )
+        len_S = len_batch * len_s
+        len_R = len_r * len_c
+
+        TAG_S, TAG_R = "threadIdx.y", "threadIdx.x"
         if target.kind.name == "cuda":
```

Comment:
The dlight folder structure assumes `dlight.gpu` for generic GPU schedules, while `dlight.cuda` for cuda-specific ones. Is it possible to refactor the arch-specific code into their own folders accordingly, or perhaps it's not a good design?

Reply:
Maybe for GEMM it makes sense to write some specific CUDA schedule, for GEMV I haven't seen any necessity here.